### DISCUSSION TITLE
✅ Unified server list for SSH and SFTP : Should or Shouldn't?

### DISCUSSION THREAD

**Discussion by josemmba** [isontheline/pro.webssh.net#185] [Ideas]
A good idea is same list for SSH and FTP. 
In this way, the same server would not have to be registered twice.

**Comment by isontheline** [comment]
Hello @josemmba👋

I agree with you but a few years ago a lot of users have asked to split these two lists in order to don't have a popup asking : "what do you want to launch ssh or sftp?"

😕

Perhaps a setting is the best answer in order to switch between the two modes? 😇

Have a nice day☀️

**Reply by josemmba** [reply]
Another option is to present the Server and two buttons, SSH and SFTP, or a long press on the name of the server and a pop-up window appears with the two options.

**Reply by isontheline** [reply]
I'm not convinced about these options :
* buttons : won't be user/Apple friendly I think
* long press : how a user will know that?

A poll for this feature could be great! 
Because I want to implement the best user interface for everyone

**Reply by josemmba** [reply]
Other option . If you add SSH server, automatically add the same server in SFTP tab.

**Reply by isontheline** [reply]
Yes it could.

I'm opening an issue to track this new feature : https://github.com/isontheline/pro.webssh.net/issues/186

**Comment by CapableFish** [comment]
Why don't you add 2 buttons next to the name of the server, one buttons for SFTP and another button for SSH

**Reply by isontheline** [reply]
@josemmba have already asked for two buttons : https://github.com/isontheline/pro.webssh.net/discussions/185#discussioncomment-357773

As you are now at least two users with the same request I'm opening an issue : https://github.com/isontheline/pro.webssh.net/issues/202

**Reply by cybardev** [reply]
Another solution could be a "toggle" switch in the top HUD to switch between the two modes, since it's just binary modes.

**Reply by isontheline** [reply]
Hello @cybardev 👋

Thanks for your feedback!

The toggle may only go inside settings because there will be more "top HUD" buttons in the following releases.

Moreover the "unified server list" should be the only one way to go at some point.

**Reply by cybardev** [reply]
Ah, that makes sense. Thanks for the info. Looking forward to what's in store for the app 😊

**Comment by BobDenny** [comment]
FYI I like it as is. Clone the SSH entry as SFTP, done. My lists are already long. The button idea may be ok though.

**Reply by isontheline** [reply]
Hello @BobDenny 👋

Thank you for your feedback 🙏

I'm [planning](https://github.com/isontheline/pro.webssh.net/issues/202) to add a dropdown "capabilities" inside connections :
* SSH
* SFTP
* SCP
* ESXi
* Docker
* PVE
* ...

So only one connection will be created for multiple protocols if they stand for the same host.

What do you think about?

**Reply by BobDenny** [reply]
That would indeed be good!!

**Comment by bkonia** [comment]
I don't understand why we have to create separate connections for SSH, SFTP and tunnels. This is wasteful of resources and redundant. Instead, they should all run over the same connection. The idea is to create one SSH connection to a server. Once the connection is established, the user can toggle between screens for terminal, file transfer and tunneling. These three "views" could be displayed as sub-tabs on a row below the active tab. You could also provide the ability to disable one or more of the views if the user doesn't want to use them. For example, there could be two sub-tabs for just terminal and file transfer. Or, if you just want a terminal session, it would be displayed in the top-level tab without any sub-tabs.

**Reply by isontheline** [reply]
> Once the connection is established, the user can toggle between screens for terminal, file transfer and tunneling.

I think port forwarding will not be a child of a connection / server. 
It should remain independent as sometimes we need to launch tunnels before launching a SSH or SFTP connection.

**Comment by isontheline** [comment]
@bkonia You're right the chosen implementation is wrong.

An issue already exists : https://github.com/isontheline/pro.webssh.net/issues/202
I'm adding your comment to this issue.

I can't give any ETA yet.
